sound/soc/sof/topology.c

```diff
@@ -2477,34 +2479,82 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
 int snd_sof_load_topology(struct snd_soc_component *scomp, const char *file)
 {
 	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);
+	struct snd_sof_pdata *sof_pdata = sdev->pdata;
 	const struct firmware *fw;
+	const char **tplg_files;
+	int tplg_cnt = 0;
 	int ret;
+	int i;
 
-	dev_dbg(scomp->dev, "loading topology:%s\n", file);
+	tplg_files = kcalloc(scomp->card->num_links, sizeof(char *), GFP_KERNEL);
+	if (!tplg_files)
+		return -ENOMEM;
 
-	ret = request_firmware(&fw, file, scomp->dev);
-	if (ret < 0) {
-		dev_err(scomp->dev, "error: tplg request firmware %s failed err: %d\n",
-			file, ret);
-		dev_err(scomp->dev,
-			"you may need to download the firmware from https://github.com/thesofproject/sof-bin/\n");
-		return ret;
+	if (sof_pdata->machine->get_tplg_files) {
+		tplg_cnt = sof_pdata->machine->get_tplg_files(scomp->card, sof_pdata->machine,
+							      sof_pdata->tplg_filename_prefix,
+							      &tplg_files);
+		if (tplg_cnt < 0) {
+			kfree(tplg_files);
+			return tplg_cnt;
+		}
 	}
 
-	if (sdev->dspless_mode_selected)
-		ret = snd_soc_tplg_component_load(scomp, &sof_dspless_tplg_ops, fw);
-	else
-		ret = snd_soc_tplg_component_load(scomp, &sof_tplg_ops, fw);
+	if (!tplg_cnt) {
+		tplg_files[0] = file;
+		tplg_cnt = 1;
+	}
 
-	if (ret < 0)
-		dev_err(scomp->dev, "error: tplg component load failed %d\n",
-			ret);
+	for (i = 0; i < tplg_cnt; i++) {
+		dev_dbg(scomp->dev, "loading topology %d: %s\n", i, tplg_files[i]);
+		ret = firmware_request_nowarn(&fw, tplg_files[i], scomp->dev);
```

Comment:
I think the get_tplg_files() should be checking the presence of the files or have a separate loop which only does that.
If any of the files are missing then fall-back to the combined file, which is known to exist as it was checked earlier during probe.

Reply:
Ok, I will do the change.